### PR TITLE
send mavlink Mission item reached

### PR DESCRIPTION
**Pull request by magicrub**
Send mavlink msg mission_item_reached when a mission item has completed.

fixes https://github.com/diydrones/ardupilot/issues/2169

### CODE REVIEW

**Comment by WickedShell** [comment]
Looks good to me. (Haven't tried it out, just read through it).

**Comment by magicrub** [comment]
I have not tested this, it's a copy/paste from @rmackay9 solution for copter. Seems simple enough though.

**Comment by tridge** [comment]
this needs a check that there is enough buffer space on the channel before sending. If we don't check for available buffer space and there isn't enough then we end up losing both this message and potentially the next message (as the GCS tries to resync the stream). If there isn't enough buffer space it is better to just not send the message.
I also think we could move the looping logic into libraries/GCS_MAVLink, similar to send_statustext_all()

**Comment by magicrub** [comment]
redesigning to use message try queue. Will resubmit rover, copter and plane under a new single PR

**Comment by magicrub** [comment]
Ok, I've completely re-wrote this so it uses the try_send que method. it also has a static variable to capture the correct index at time of sending initial index. If it was to try to resend later the index would have changed so this was required in some form

**Comment by magicrub** [comment]
Changes:
- rebased on master
- moved more code into GCS_Common
- replaced static index with class member
- unified across all 4 platforms

TODO: As soon as RAM is not a critical as it is now for APM 2.x platforms, I propose we change the "try" stuff to use a circular buffer which is a class that includes a param. That will ensure all msgs get sent as you wish them to get sent and none will get lost when additional msgs come in of the same type

**Comment by rmackay9** [comment]
I'm not sure if it's critical or not but this new method will not send a message when a do-command is completed like the current method implemented in master for copter.

**Comment by magicrub** [comment]
It would need to report reached of all mission item types to be correct.
I'll take a look.

File: APMrover2/commands_logic.cpp [hunk 1]
```diff
@@ -122,6 +122,25 @@ void Rover::exit_mission()
...
+
+        // send message to GCS
+        if (cmd_complete) {
+            gcs_send_message(MSG_MISSION_ITEM_REACHED);
+        }
+
+        return cmd_complete;
+    }
+
+    // exit immediately if not in AUTO mode
```
Comment (tridge) [inline APMrover2/commands_logic.cpp]: copter and plane return false here. Maybe it doesn't matter?

**Comment by tridge** [comment]
I think there is a problem where if there are deferred messages at the time that gcs_send_message(MSG_MISSION_ITEM_REACHED) is called, and the serial port is full, then it will end up sending a MISSION_ITEM_REACHED with the next wp number, not the one that has just been reached.
I think the fix is to set _mission_item_reached_index at the time gcs_send_message() is called, and not rely on it being set by try_send_message(). The try_send_message() function is not guaranteed to be called immediately, and if it isn't then the current wp in the mission object will have moved on, and we'll end up sending a message for the wrong wp

File: APMrover2/commands_logic.cpp (hunk 1 above)
Comment (magicrub) [inline APMrover2/commands_logic.cpp]: agreed, all should return false. However, as Randy as noted this callback can't even get called when not in auto so the whole control_mode == AUTO check, and the else return false are useless code. If we want to remove any of that lets do it in a separate PR

**Comment by magicrub** [comment]
@randy and @tridge your suggestions have been implemented. the index is stored at the callback time. Also rover is changed to "false". I rebase interactively the plane changes but left rover and copter changes ad additions so its easier to see what I changed

**Comment by magicrub** [comment]
@rmackay9 I tested this with some DO_SET_SERVO commands and they reported as expected but the DO_JUMP doesn't and thats because there's nothings actually executing to trigger a reached so I'm ok with that.

**Comment by magicrub** [comment]
BUMP - rebased on master

**Comment by tridge** [comment]
applied with minor changes as discussed
thanks!
